angr/analyses/decompiler/block_io_finder.py

```diff
@@ -0,0 +1,292 @@
+from ailment.expression import (
+    Load,
+    Expression,
+    BinaryOp,
+    UnaryOp,
+    Convert,
+    ITE,
+    Tmp,
+    Const,
+    StackBaseOffset,
+)
+from ailment.block_walker import AILBlockWalkerBase
+
+
+from angr.knowledge_plugins.key_definitions.atoms import MemoryLocation, Register, SpOffset, ConstantSrc
+
+
+class BlockIOFinder(AILBlockWalkerBase):
+    """
+    Finds the input and output locations of each statement in an AIL block.
+    I/O locations can be a Register, MemoryLocation, or SpOffset (wrapped in a Memory Location).
+    """
+
+    def __init__(self, ail_obj: Union[Block, List[Statement]], project, as_atom=True):
+        super().__init__()
+        self.expr_handlers[StackBaseOffset] = self._handle_StackBaseOffset
+        self._as_atom = as_atom
+        self._project = project
+
+        self.inputs_by_stmt = defaultdict(set)
+        self.outputs_by_stmt = defaultdict(set)
+        self.derefed_at = defaultdict(set)
+
+        block = Block(0, len(ail_obj), statements=ail_obj) if isinstance(ail_obj, list) else ail_obj
+        self.walk(block)
+
+    @staticmethod
+    def _add_or_update_dict(d, k, v):
+        if isinstance(v, set):
+            d[k].update(v)
+        else:
+            d[k].add(v)
+
+    @staticmethod
+    def _add_or_update_set(s, v):
+        if isinstance(v, set):
+            s.update(v)
+        else:
+            s.add(v)
+
+    #
+    # I/O helpers
+    #
+
+    @staticmethod
+    def _is_dangerous_memory(loc):
+        """
+        Assume any memory location that is NOT on the stack is a dangerous memory location.
+        """
+        return isinstance(loc, MemoryLocation) and not loc.is_on_stack
+
+    def _has_dangerous_deref(self, stmt_idx):
+        derefs = self.derefed_at.get(stmt_idx, set())
+        return any(self._is_dangerous_memory(d) for d in derefs)
+
+    def _input_defined_by_other_stmt(self, target_idx, other_idx):
+        target_inputs = self.inputs_by_stmt[target_idx]
+        # any memory location, not on stack, is not movable
+        if any(self._is_dangerous_memory(i) for i in target_inputs):
+            return True
+
+        other_outputs = self.outputs_by_stmt[other_idx]
+        return target_inputs.intersection(other_outputs)
+
+    def _output_used_by_other_stmt(self, target_idx, other_idx):
+        target_output = self.outputs_by_stmt[target_idx]
+        # any memory location, not on stack, is not movable
+        if any(self._is_dangerous_memory(o) for o in target_output):
+            return True
+
+        other_input = self.inputs_by_stmt[other_idx]
+        return target_output.intersection(other_input)
+
+    def can_swap(self, stmt, ail_obj: Union[Block, List[Statement]], offset: int):
+        all_stmts = (ail_obj.statements or []) if isinstance(ail_obj, Block) else ail_obj
+        if stmt not in all_stmts:
+            raise NotImplementedError("Statement not in block, and we can't compute moving a stmt to a new block!")
```

Comment:
`NotImplementedError` means you'll implement it in the future. I suppose that's not what you mean here?